src/workerd/api/worker-rpc.c++

```diff
@@ -200,39 +260,107 @@ private:
+    auto f = kj::mv(JSG_REQUIRE_NONNULL(doneFulfiller,
+        Error, "Only one RPC method call is allowed on this object."));
+    doneFulfiller = kj::none;
+    return capnp::membrane(kj::mv(target), kj::refcounted<CompletionMembrane>(kj::mv(f)));
+  }
+
+  kj::Maybe<capnp::Capability::Client> outboundCall(
+      uint64_t interfaceId, uint16_t methodId, capnp::Capability::Client target) override {
+    KJ_FAIL_ASSERT("ServerTopLevelMembrane shouldn't have outgoing capabilities");
+  }
+
+  kj::Own<MembranePolicy> addRef() override {
+    return kj::addRef(*this);
+  }
+
+private:
+  kj::Maybe<kj::Own<kj::PromiseFulfiller<void>>> doneFulfiller;
+};
+
+kj::Promise<WorkerInterface::CustomEvent::Result> JsRpcSessionCustomEventImpl::run(
     kj::Own<IoContext::IncomingRequest> incomingRequest,
     kj::Maybe<kj::StringPtr> entrypointName) {
   incomingRequest->delivered();
-  auto [callPromise, callFulfiller] = kj::newPromiseAndFulfiller<void>();
-  capFulfiller->fulfill(kj::heap<JsRpcTargetImpl>(
-      kj::mv(callFulfiller), incomingRequest->getContext(), entrypointName));
-
-  // `callPromise` resolves once `JsRpcTargetImpl::call()` (invoked by client) completes.
-  co_await callPromise;
+  auto [donePromise, doneFulfiller] = kj::newPromiseAndFulfiller<void>();
+  capFulfiller->fulfill(
+      capnp::membrane(
+          kj::heap<JsRpcTargetImpl>(incomingRequest->getContext(), entrypointName),
+          kj::refcounted<ServerTopLevelMembrane>(kj::mv(doneFulfiller))));
+
+  // `donePromise` resolves once there are no longer any capabilities pointing between the client
+  // and server as part of this session.
+  co_await donePromise;
   co_await incomingRequest->drain();
   co_return WorkerInterface::CustomEvent::Result {
     .outcome = EventOutcome::OK
   };
 }
 
 kj::Promise<WorkerInterface::CustomEvent::Result>
-  GetJsRpcTargetCustomEventImpl::sendRpc(
+  JsRpcSessionCustomEventImpl::sendRpc(
     capnp::HttpOverCapnpFactory& httpOverCapnpFactory,
     capnp::ByteStreamFactory& byteStreamFactory,
     kj::TaskSet& waitUntilTasks,
     rpc::EventDispatcher::Client dispatcher) {
-  auto req = dispatcher.getJsRpcTargetRequest();
+  // We arrange to revoke all capabilities in this session as soon as `sendRpc()` completes or is
+  // canceled. Normally, the server side doesn't return if any capabilities still exist, so this
+  // only makes a difference in the case that some sort of an error occurred. We don't strictly
+  // have to revoke the capabilities as they are probably already broken anyway, but revoking them
+  // helps to ensure that the underlying transport isn't "held open" waiting for the JS garbage
+  // collector to actually collect the JsRpcCapability objects.
+  auto [revokePromise, revokeFulfiller] = kj::newPromiseAndFulfiller<void>();
+
+  KJ_DEFER({
+    if (revokeFulfiller->isWaiting()) {
+      revokeFulfiller->reject(KJ_EXCEPTION(DISCONNECTED, "JS-RPC session canceled"));
+    }
```

Comment:
Compiler in CI doesn't like this for some reason. If I'm recalling correctly, I think there was an old clang bug with capturing a reference to a structured binding, can't remember the details tho (https://stackoverflow.com/questions/46114214/lambda-implicit-capture-fails-with-variable-declared-from-structured-binding)

Reply:
Huh, I guess this "bug in the standard" must have been fixed in the newer clang we're using for our internal build, but not in the default clang used in CI? Fixed by removing the destructuring.